File: telemetry-service/src/main/java/net/blueshell/telemetry/repository/RedirectRepository.java
package net.blueshell.telemetry.repository;

import jakarta.validation.constraints.NotNull;
import net.blueshell.db.BaseRepository;
import net.blueshell.telemetry.model.Redirect;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

@Repository
public interface RedirectRepository extends BaseRepository<Redirect, UUID> {

    @Query("SELECT n FROM Redirect n ORDER BY n.createdAt DESC")
    @Override
    Page<Redirect> findAll(@NotNull Pageable pageable);

    @Query("SELECT n FROM Redirect n ORDER BY n.createdAt DESC")
    @Override
    List<Redirect> findAll();

    @Query("SELECT e FROM Redirect e WHERE e.createdAt >= :from AND e.createdAt <= :to ORDER BY e.createdAt DESC")
    List<Redirect>  findCreatedAtBetween(OffsetDateTime from, OffsetDateTime to);
}

File: api/src/main/kotlin/net/blueshell/api/domain/telemetry/persistence/repository/RedirectRepository.kt
package net.blueshell.api.domain.telemetry.persistence.repository

import jakarta.validation.constraints.NotNull
import net.blueshell.api.shared.repository.BaseRepository
import net.blueshell.api.telemetry.persistence.Redirect
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import java.time.OffsetDateTime

@Repository
interface RedirectRepository : BaseRepository<Redirect, Long> {
    @Query("SELECT n FROM Redirect n ORDER BY n.createdAt DESC")
    override fun findAll(@NotNull pageable: @NotNull Pageable): Page<Redirect>

    @Query("SELECT n FROM Redirect n ORDER BY n.createdAt DESC")
    override fun findAll(): MutableList<Redirect>

    @Query("SELECT e FROM Redirect e WHERE e.createdAt >= :from AND e.createdAt <= :to ORDER BY e.createdAt DESC")
    fun findCreatedAtBetween(from: OffsetDateTime, to: OffsetDateTime): MutableList<Redirect>
}
